diff -uwpr ../../../../GeoIP-1.4.8/libGeoIP/GeoIP.h ./GeoIP.h
--- ../../../../GeoIP-1.4.8/libGeoIP/GeoIP.h	2011-06-23 22:57:02.000000000 +0200
+++ ./GeoIP.h	2013-02-23 17:48:36.607250000 +0100
@@ -33,7 +33,7 @@ extern "C" {
 #else /* !defined(_WIN32) */ 
 #include <winsock2.h> 
 #include <ws2tcpip.h> 
-#include <wspiapi.h>
+//#include <wspiapi.h>
 #include <windows.h> 
 #define snprintf _snprintf 
 #define FILETIME_TO_USEC(ft) (((unsigned __int64) ft.dwHighDateTime << 32 | ft.dwLowDateTime) / 10) 
diff -uwpr ../../../../GeoIP-1.4.8/libGeoIP/GeoIPCity.c ./GeoIPCity.c
--- ../../../../GeoIP-1.4.8/libGeoIP/GeoIPCity.c	2011-03-30 19:55:14.000000000 +0200
+++ ./GeoIPCity.c	2013-02-23 17:57:02.716625000 +0100
@@ -38,8 +38,8 @@
 static
 const int       FULL_RECORD_LENGTH = 50;
 
-static const int CITYCONFIDENCE_FIXED_RECORD = 4;
-static const int CITYCONFIDENCEDIST_FIXED_RECORD = 6;
+#define CITYCONFIDENCE_FIXED_RECORD      4
+#define CITYCONFIDENCEDIST_FIXED_RECORD  6
 
 
 static
@@ -55,6 +55,7 @@ _extract_record(GeoIP * gi, unsigned int
   double          latitude = 0, longitude = 0;
   int             metroarea_combo = 0;
   int             bytes_read = 0;
+  int             t = 0;
   if (seek_record == gi->databaseSegments[0])
     return NULL;
 
@@ -97,7 +98,7 @@ _extract_record(GeoIP * gi, unsigned int
 	gi->databaseType == GEOIP_CITYCONFIDENCEDIST_EDITION
 	? ((tmp_fixed_record[4] + (tmp_fixed_record[5] << 8)) & 0x3ff) : 0x3ff;
 
-      int             t = fixed_rec_size - gi->record_length;
+      t = fixed_rec_size - gi->record_length;
       
     record_pointer = dseg + tmp_fixed_record[t] +
 	(tmp_fixed_record[t + 1] << 8) + (tmp_fixed_record[t + 2] << 16) ;
@@ -129,7 +130,7 @@ _extract_record(GeoIP * gi, unsigned int
 	gi->databaseType == GEOIP_CITYCONFIDENCEDIST_EDITION
 	? ((record_buf[4] + (record_buf[5] << 8)) & 0x3ff) : 0x3ff;
 
-      int             t = fixed_rec_size - gi->record_length;
+      t = fixed_rec_size - gi->record_length;
 
         record_pointer = dseg + record_buf[t] +
 	(record_buf[t + 1] << 8) + (record_buf[t + 2] << 16) ;
@@ -210,13 +211,13 @@ _extract_record(GeoIP * gi, unsigned int
   /* get latitude */
   for (j = 0; j < 3; ++j)
     latitude += (record_buf[j] << (j * 8));
-  record->latitude = latitude / 10000 - 180;
+  record->latitude = (float)(latitude / 10000 - 180);
   record_buf += 3;
 
   /* get longitude */
   for (j = 0; j < 3; ++j)
     longitude += (record_buf[j] << (j * 8));
-  record->longitude = longitude / 10000 - 180;
+  record->longitude = (float)(longitude / 10000 - 180);
 
   /*
    * get area code and metro code for post April 2002 databases and for US
